internal/driver/glfw/window.go

```diff
@@ -256,21 +251,29 @@ func (w *window) fitContent() {
 	minWidth, minHeight := w.minSizeOnScreen()
 	w.viewLock.Lock()
 	defer w.viewLock.Unlock()
-	if w.width < minWidth || w.height < minHeight {
-		if w.width < minWidth {
-			w.width = minWidth
-		}
-		if w.height < minHeight {
-			w.height = minHeight
-		}
-		w.viewport.SetSize(w.width, w.height)
-	}
 	if w.fixedSize {
-		w.width = internal.ScaleInt(w.canvas, w.Canvas().Size().Width)
-		w.height = internal.ScaleInt(w.canvas, w.Canvas().Size().Height)
+		limitSize := internal.ScaleSize(w.canvas, w.Canvas().Size())
+		if limitSize.Width == 0 || limitSize.Height == 0 {
+			// When the content hasn't been set yet, the min size is not available yet and
+			// their is no point into doing anything
+			return
+		}
 
-		w.viewport.SetSizeLimits(w.width, w.height, w.width, w.height)
+		w.viewport.SetSizeLimits(limitSize.Width, limitSize.Height, limitSize.Width, limitSize.Height)
+		w.viewport.SetSize(limitSize.Width, limitSize.Height)
 	} else {
+		// Adjust size to be at least minSize
+		if w.size.Width < minWidth || w.size.Height < minHeight {
+			targetWidth, targetHeight := w.size.Width, w.size.Height
+			if targetWidth < minWidth {
+				targetWidth = minWidth
+			}
+			if targetHeight < minHeight {
+				targetHeight = minHeight
+			}
+			w.viewport.SetSize(targetWidth, targetHeight)
```

Comment:
In the end `fitContent` now changes the viewport size without adjusting the window’s size as well. This feels wrong. Maybe some other code adjusts this later but I think the window’s size and the viewport size should be very thightly coupled. A method should not change one without changing the other.

Reply:
If `viewport.GetSize` is cheap, we could drop the window’s size property at all.

Reply:
The viewport size should not be adjusted outside the render loop I think - this may be causing some of the resize flicker?

Reply:
I think you are missing some context. When you do call viewport.SetSize, the outcome of that call is not that the window is resized once that call is done, but that a request has been made to resize the window. The window resize call is asynchronous in nature and any change can be denied by the window manager. So you can only update the window size once you get the callback from the window manager that tell your new window size. And any point in between your window is still the same size as it was when you initiated the viewport.SetSize.

Basically the entire point of this patch set is to make sure that the viewport size and the window size always reflect the true value that the window manager has given it at any point in time.

Reply:
Also I would not assume that the viewport.GetSize is cheap as on some system it might require a round with the window manager to get that information. I do not know if GLFW does cache that value for us.

Reply:
Oops, I was confusing viewport.SetSize and gl.Viewport() sorry!

Reply:
Thanks for the explanation. Then of course my comment is totally wrong. Alas, they could have named the method `RequestResize` :).